File: library/src/main/java/org/meganekkovr/xml/RotationHandler.java
package org.meganekkovr.xml;

import android.content.Context;

import org.joml.Quaternionf;
import org.meganekkovr.Entity;

import java.util.Locale;

/**
 * Define {@code rotation} attribute.
 */
public class RotationHandler implements XmlAttributeParser.XmlAttributeHandler {

    @Override
    public String attributeName() {
        return "rotation";
    }

    @Override
    public void parse(Entity entity, String rawValue, Context context) {

        String[] strs = rawValue.split("\\s+", 3);
        if (strs.length >= 3) {

            float x = (float) Math.toRadians(Float.parseFloat(strs[0]));
            float y = (float) Math.toRadians(Float.parseFloat(strs[1]));
            float z = (float) Math.toRadians(Float.parseFloat(strs[2]));

            String rotOrder = strs.length == 4 ? strs[3].toUpperCase(Locale.US) : "XYZ";
            Quaternionf q = new Quaternionf();
            switch (rotOrder) {
                case "XYZ":
                    q.rotateXYZ(x, y, z);
                    break;
                case "YXZ":
                    q.rotateYXZ(x, y, z);
                    break;
                case "ZYX":
                    q.rotateZYX(x, y, z);
                    break;
            }
            entity.setRotation(q);
        }
    }
}

File: library/src/main/java/org/meganekkovr/xml/RotationHandler.kt
package org.meganekkovr.xml

import android.content.Context
import org.joml.Quaternionf
import org.meganekkovr.Entity
import java.util.*

/**
 * Define `rotation` attribute.
 */
internal class RotationHandler : XmlAttributeParser.XmlAttributeHandler {

    override fun attributeName(): String {
        return "rotation"
    }

    override fun parse(entity: Entity, rawValue: String, context: Context) {

        val strs = rawValue.split("\\s+".toRegex(), 3)
        if (strs.size >= 3) {

            val x = strs[0].toDoubleOrNull() ?: return
            val y = strs[1].toDoubleOrNull() ?: return
            val z = strs[2].toDoubleOrNull() ?: return

            val xRad = Math.toRadians(x).toFloat()
            val yRad = Math.toRadians(y).toFloat()
            val zRad = Math.toRadians(z).toFloat()

            val rotOrder = if (strs.size == 4) strs[3].toUpperCase(Locale.US) else "XYZ"
            val q = Quaternionf()
            when (rotOrder) {
                "XYZ" -> q.rotateXYZ(xRad, yRad, zRad)
                "YXZ" -> q.rotateYXZ(xRad, yRad, zRad)
                "ZYX" -> q.rotateZYX(xRad, yRad, zRad)
            }
            entity.rotation = q
        }
    }
}
